src/ephy-sync-service.c

```diff
@@ -0,0 +1,1417 @@
+    storage_server_request_async_data_free (data);
+    service->locked = FALSE;
+    goto out;
+  }
+
+  service->certificate = g_strdup (certificate);
+
+  /* See the comment in ephy_sync_service_send_storage_message(). */
+  ephy_sync_service_obtain_storage_credentials (service, user_data);
+
+out:
+  g_object_unref (parser);
+}
+
+static void
+ephy_sync_service_obtain_signed_certificate (EphySyncService *self,
+                                             gpointer         user_data)
+{
+  guint8 *tokenID;
+  guint8 *reqHMACkey;
+  guint8 *requestKey;
+  char *tokenID_hex;
+  char *public_key_json;
+  char *request_body;
+  char *n;
+  char *e;
+
+  g_return_if_fail (EPHY_IS_SYNC_SERVICE (self));
+  g_return_if_fail (self->sessionToken != NULL);
+
+  if (self->keypair != NULL)
+    ephy_sync_crypto_rsa_key_pair_free (self->keypair);
+
+  self->keypair = ephy_sync_crypto_generate_rsa_key_pair ();
+  g_return_if_fail (self->keypair != NULL);
+
+  ephy_sync_crypto_process_session_token (self->sessionToken, &tokenID, &reqHMACkey, &requestKey);
+  tokenID_hex = ephy_sync_crypto_encode_hex (tokenID, 0);
+
+  n = mpz_get_str (NULL, 10, self->keypair->public.n);
+  e = mpz_get_str (NULL, 10, self->keypair->public.e);
+  public_key_json = ephy_sync_utils_build_json_string ("algorithm", "RS", "n", n, "e", e, NULL);
+  /* Duration is the lifetime of the certificate in milliseconds. The FxA server
+   * limits the duration to 24 hours. For our purposes, a duration of 30 minutes
+   * will suffice.
+   */
+  request_body = g_strdup_printf ("{\"publicKey\": %s, \"duration\": %d}",
+                                  public_key_json, 30 * 60 * 1000);
+  ephy_sync_service_fxa_hawk_post_async (self, "certificate/sign", tokenID_hex,
+                                         reqHMACkey, EPHY_SYNC_TOKEN_LENGTH, request_body,
+                                         obtain_signed_certificate_response_cb, user_data);
+
+  g_free (tokenID);
+  g_free (reqHMACkey);
+  g_free (requestKey);
+  g_free (tokenID_hex);
+  g_free (public_key_json);
+  g_free (request_body);
+  g_free (n);
+  g_free (e);
+}
+
+static void
+ephy_sync_service_issue_storage_request (EphySyncService               *self,
+                                         StorageServerRequestAsyncData *data)
+{
+  g_return_if_fail (EPHY_IS_SYNC_SERVICE (self));
+  g_return_if_fail (data != NULL);
+
+  if (ephy_sync_service_storage_credentials_is_expired (self) == TRUE) {
+    ephy_sync_service_clear_storage_credentials (self);
+
+    /* The only purpose of certificates is to obtain a signed BrowserID that is
+     * needed to talk to the Token Server. From the Token Server we will obtain
+     * the credentials needed to talk to the Storage Server. Since both
+     * ephy_sync_service_obtain_signed_certificate() and
+     * ephy_sync_service_obtain_storage_credentials() complete asynchronously,
+     * we need to entrust them the task of sending the request to the Storage
```

Comment:
I'm not sure this is the cleanest solution, actually. Your functions should do just the things their names say they will do. You can pass callbacks to be executed asynchronously once they've completed.